### PR TITLE
Create computational scaling plots with py-test benchmark and post to github actions artifact

### PR DESCRIPTION
**Pull request by bparazin**
At the moment I am just getting started on seeing if you can use pytest-benchmark to produce computational scaling plots, and to that end, I'm making a PR so I can test functionality things. Don't expect this to be merged any time soon.

### CODE REVIEW

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -49,4 +49,15 @@ jobs:
         poetry install
     - name: Test with pytest
       run: |
-        poetry run pytest --benchmark-sort=fullname
+        poetry run pytest --benchmark-sort=fullname --benchmark-json output.json
```
Comment (lpsinger) [inline .github/workflows/build-and-test.yml:52]: Please name the output `benchmark.json` to be a little bit more descriptive.

File: .github/workflows/build-and-test.yml [hunk 2]
```diff
@@ -49,4 +49,15 @@ jobs:
...
+      if: ${{ matrix.os == 'ubuntu-latest' && matrix.python-version == '3.9'}}
+      run: |
+        mkdir plots
+        poetry run python .github/workflows/runtime_plots.py
+    - name: Upload plots to github artifacts
+      if: ${{ matrix.os == 'ubuntu-latest' && matrix.python-version == '3.9'}}
+      uses: actions/upload-artifact@v2
+      with:
+        name: runtime-plots
+        path: plots
```
Comment (lpsinger) [inline .github/workflows/build-and-test.yml:63]: Fix missing newline at end of file.

Comment (lpsinger) [inline .github/workflows/build-and-test.yml:63]: Also upload the json file.

File: .github/workflows/runtime_plots.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+def data_list_to_dict(data):
+    result_dict = {}
+    for test_run in data['benchmarks']:
+        test = test_run['fullname'].split('[')[0]
+        if test in result_dict.keys():
+            result_dict[test].append(test_run)
+        else:
+            result_dict[test] = [test_run]
+    return result_dict
```
Comment (lpsinger) [inline .github/workflows/runtime_plots.py:15]: Here's a slick way to read the data file using pandas:

```pycon
>>> import json
>>> import pandas
>>> with open('benchmark.json') as f:
...     data = json.load(f)
... 
>>> df = pandas.json_normalize(data['benchmarks'])
>>> df
   group                             name                                           fullname  param  params.point_clouds  options.disable_gc  ... stats.outliers  stats.ld15iqr  stats.hd15iqr   stats.ops  stats.total  stats.iterations
0   None             test_cross_join[100]  healpix_alchemy/tests/test_point.py::test_cros...    100                  100               False  ...          59;22       0.001806       0.002652  449.213747     0.605502                 1
1   None  test_cone_search_literal[10000]  healpix_alchemy/tests/test_point.py::test_cone...  10000                10000               False  ...          77;34       0.001604       0.002292  523.021912     0.650068                 1
2   None          test_cone_search[10000]  healpix_alchemy/tests/test_point.py::test_cone...  10000                10000               False  ...          84;36       0.001599       0.002234  525.237350     0.795831                 1
3   None           test_cross_join[10000]  healpix_alchemy/tests/test_point.py::test_cros...  10000                10000               False  ...            1;0       0.542646       0.590758    1.794740     2.785920                 1
4   None             test_cone_search[10]  healpix_alchemy/tests/test_point.py::test_cone...     10                   10               False  ...          56;37       0.001530       0.002303  536.787653     0.817828                 1
5   None     test_cone_search_literal[10]  healpix_alchemy/tests/test_point.py::test_cone...     10                   10               False  ...          62;26       0.001526       0.002328  548.148538     0.738851                 1
6   None            test_cross_join[1000]  healpix_alchemy/tests/test_point.py::test_cros...   1000                 1000               False  ...           27;1       0.010859       0.014313   83.251695     0.960941                 1
7   None    test_cone_search_literal[100]  healpix_alchemy/tests/test_point.py::test_cone...    100                  100               False  ...          33;29       0.001520       0.001972  582.894899     0.679368                 1
8   None              test_self_join[100]  healpix_alchemy/tests/test_point.py::test_self...    100                  100               False  ...          39;10       0.002642       0.003738  329.810803     0.670081                 1
9   None            test_self_join[10000]  healpix_alchemy/tests/test_point.py::test_self...  10000                10000               False  ...            1;0       0.539789       0.595685    1.783983     2.802718                 1
10  None   test_cone_search_literal[1000]  healpix_alchemy/tests/test_point.py::test_cone...   1000                 1000               False  ...           1;24       0.001541       0.001972  545.907725     0.646629                 1
11  None             test_self_join[1000]  healpix_alchemy/tests/test_point.py::test_self...   1000                 1000               False  ...            1;2       0.013652       0.018410   65.601229     0.960348                 1
12  None            test_cone_search[100]  healpix_alchemy/tests/test_point.py::test_cone...    100                  100               False  ...          35;31       0.001540       0.002041  568.720541     0.754325                 1
13  None           test_cone_search[1000]  healpix_alchemy/tests/test_point.py::test_cone...   1000                 1000               False  ...          33;29       0.001532       0.002055  572.133572     0.713120                 1
14  None               test_self_join[10]  healpix_alchemy/tests/test_point.py::test_self...     10                   10               False  ...           1;22       0.002146       0.003021  376.334905     0.741361                 1
15  None              test_cross_join[10]  healpix_alchemy/tests/test_point.py::test_cros...     10                   10               False  ...          32;29       0.001649       0.002087  548.309833     0.676625                 1

[16 rows x 28 columns]
```

You can strip off the parameters from the function names like this:

```pycon
>>> df['func'] = df['name'].str.split('[', expand=True)[0]
>>> df['fullfunc'] = df['fullname'].str.split('[', expand=True)[0]
>>> df
   group                             name                                           fullname  param  params.point_clouds  ...   stats.ops stats.total  stats.iterations                      func                                           fullfunc
0   None             test_cross_join[100]  healpix_alchemy/tests/test_point.py::test_cros...    100                  100  ...  449.213747    0.605502                 1           test_cross_join  healpix_alchemy/tests/test_point.py::test_cros...
1   None  test_cone_search_literal[10000]  healpix_alchemy/tests/test_point.py::test_cone...  10000                10000  ...  523.021912    0.650068                 1  test_cone_search_literal  healpix_alchemy/tests/test_point.py::test_cone...
2   None          test_cone_search[10000]  healpix_alchemy/tests/test_point.py::test_cone...  10000                10000  ...  525.237350    0.795831                 1          test_cone_search  healpix_alchemy/tests/test_point.py::test_cone...
3   None           test_cross_join[10000]  healpix_alchemy/tests/test_point.py::test_cros...  10000                10000  ...    1.794740    2.785920                 1           test_cross_join  healpix_alchemy/tests/test_point.py::test_cros...
4   None             test_cone_search[10]  healpix_alchemy/tests/test_point.py::test_cone...     10                   10  ...  536.787653    0.817828                 1          test_cone_search  healpix_alchemy/tests/test_point.py::test_cone...
5   None     test_cone_search_literal[10]  healpix_alchemy/tests/test_point.py::test_cone...     10                   10  ...  548.148538    0.738851                 1  test_cone_search_literal  healpix_alchemy/tests/test_point.py::test_cone...
6   None            test_cross_join[1000]  healpix_alchemy/tests/test_point.py::test_cros...   1000                 1000  ...   83.251695    0.960941                 1           test_cross_join  healpix_alchemy/tests/test_point.py::test_cros...
7   None    test_cone_search_literal[100]  healpix_alchemy/tests/test_point.py::test_cone...    100                  100  ...  582.894899    0.679368                 1  test_cone_search_literal  healpix_alchemy/tests/test_point.py::test_cone...
8   None              test_self_join[100]  healpix_alchemy/tests/test_point.py::test_self...    100                  100  ...  329.810803    0.670081                 1            test_self_join  healpix_alchemy/tests/test_point.py::test_self...
9   None            test_self_join[10000]  healpix_alchemy/tests/test_point.py::test_self...  10000                10000  ...    1.783983    2.802718                 1            test_self_join  healpix_alchemy/tests/test_point.py::test_self...
10  None   test_cone_search_literal[1000]  healpix_alchemy/tests/test_point.py::test_cone...   1000                 1000  ...  545.907725    0.646629                 1  test_cone_search_literal  healpix_alchemy/tests/test_point.py::test_cone...
11  None             test_self_join[1000]  healpix_alchemy/tests/test_point.py::test_self...   1000                 1000  ...   65.601229    0.960348                 1            test_self_join  healpix_alchemy/tests/test_point.py::test_self...
12  None            test_cone_search[100]  healpix_alchemy/tests/test_point.py::test_cone...    100                  100  ...  568.720541    0.754325                 1          test_cone_search  healpix_alchemy/tests/test_point.py::test_cone...
13  None           test_cone_search[1000]  healpix_alchemy/tests/test_point.py::test_cone...   1000                 1000  ...  572.133572    0.713120                 1          test_cone_search  healpix_alchemy/tests/test_point.py::test_cone...
14  None               test_self_join[10]  healpix_alchemy/tests/test_point.py::test_self...     10                   10  ...  376.334905    0.741361                 1            test_self_join  healpix_alchemy/tests/test_point.py::test_self...
15  None              test_cross_join[10]  healpix_alchemy/tests/test_point.py::test_cros...     10                   10  ...  548.309833    0.676625                 1           test_cross_join  healpix_alchemy/tests/test_point.py::test_cros...

[16 rows x 30 columns]
```

Then you can make each plot almost in a single line with pandas:

```python
from matplotlib import pyplot as plt
for func, group in df.groupby('func'):
    fig, ax = plt.subplots()
    group.plot.scatter(x='param', y='stats.mean', yerr='stats.stddev', loglog=True, xlabel='N', ylabel='Runtime (s)', ax=ax)
    fig.tight_layout()
    fig.savefig(f'plots/{func}')
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

**Comment by bparazin** [comment]
After running into a bunch of git issues thanks to some short-sighted development on main, I have moved this to #28

**Pull request closed without merging**